### ISSUE TITLE
Resource exhausted reason is lost for job activation in gateway

### ISSUE THREAD

**Issue by korthout** [camunda/camunda#6055]
**Description**

Currently, a Zeebe broker can respond with Resource Exhausted for 2 reasons: backpressure and out of disk. A broker will clearly inform the client that the broker is out of disk space and is unable to accept requests for a particular partition. However, this information is lost when the command was an `ActivateJobs` command.

The gateway sends the `ActivateJobs` command to the brokers in a round robin fashion (even the embedded gateway does this). If one of the brokers responds with a resource exhausted another resource exhausted message is responded to the client, losing the cause of the exhaustion.

This means that it is never clear for a client whether the Resource Exhausted responds is due to backpressure, or due to out of disk space.

**Expected behaviour**

Rejection messages should contain enough relevant information for the client to know what to do or how to debug this.

**To reproduce**

1. Set `zeebe.broker.data.diskUsageCommandWatermark` to `0.01` or lower and start-up a broker. 
2. Run `zbctl --insecure publish message test --correlationKey="test"` to see a resource exhausted due to out of disk space response
```
Error: rpc error: code = ResourceExhausted desc = Cannot accept requests for partition 1. Broker is out of disk space
```
3. Run `zbctl --insecure create worker test --handler=cat` to see the follow response message:
```
2021/01/04 17:42:20 Failed to activate jobs for worker zbctl rpc error: code = ResourceExhausted desc = Expected to activate jobs of type 'test', but no jobs available and at least one broker returned 'RESOURCE_EXHAUSTED'. Please try again later.
 ```

**Comment by korthout** [comment]
`Severity: Mid` because this happens quite often

**Comment by npepinpe** [comment]
Downgraded to low, as from a pure client point of view, if you get this error, regardless of the cause, the only thing you can (and should) do is back off. SREs operating the broker/gateway will already be aware if you're out of disk space (or they should anyway) via other channels, so it doesn't strike me as severe.

I still agree we can improve the error message or at least provide more details, but low prio from my side.

**Comment by rusterholz** [comment]
Is it at all possible that the disk space message could be triggered by running out of file descriptors, rather than by actually bumping up against a byte size limit?

I ask because we see the `Failed to activate jobs` form of the ResourceExhausted error quite frequently, even when Zeebe is under very low load (appx. one workflow every five minutes), but we aren't anywhere near running out of disk space (we have 100GB allocated and the brokers only use ~10GB). We do, however, often see a broker keeping ~400,000 files open simultaneously. Is it possible that this situation could be reported by the broker as a disk space ResourceExhausted error? It shouldn't be due to backpressure as the whole cluster is under very low load, and it shouldn't be due to disk space as we are only using 10% of the available capacity, so we don't currently have a good explanation for why we're seeing this exact error so frequently.

**Comment by deepthidevaki** [comment]
@rusterholz 
400000 files on the snapshot does not cause ResourceExhausted due to out of disk space. If it is due to out of disk space, it will be logged in the broker (at info level, I guess) when out of disk space is detected for the first time. You can check the logs to verify that. 

Resource exhausted that you see should be most probably due to the normal backpressure. Even though you are only running 1 workflow per minute, if the broker is receiving many activate job requests at the same time, then backpressure can kick in. It can happen because we don't have any mechanism to handle bursts. Gateway can simultaneously release a bunch of activated job requests which are blocked due to long polling. A good way to investigate it is by looking at the metrics (prometheus + grafana).

**Comment by npepinpe** [comment]
I think you might run into this issue if you run out of inodes, as it happened to us, so having many many small files could result in this. I think @pihme investigated an issue where this occurred (correct me if I'm wrong there).

**Comment by pihme** [comment]
@rusterholz 
I investigated an issue, yes. However, when we ran out of file descriptors we did see out of disk space exceptions in the server logs, but I am not aware that we returned `RESOURCE_EXHAUSTED` to the client. 

At the call sites, where the error occurred, there was no client command to respond to. Rather the error occurred during internal replication processes.

Also, a cursory glance at the code for when we return a `RESOURCE_EXHAUSTED` response to a client indicates that we only look at free disk space, not at free file descriptors.

I might have missed some detail in the code, but from what I can see I second @deepthidevaki assessment.
